### PR TITLE
vendor: bump libnetwork 6659f7f4d8c1ec7a412a33f8973423ad38d22982

### LINKED ISSUE #30053: Unable to choose outbound (external) IP for containers
In single host mode (no swarm and more complicated stuff) I have a host with multiple public IPs. It seems there is no way to configure which of those IPs containers use for outbound communication. Always the primary IP on the host is used. So I would need that different containers are seen on the Internet as using different IPs.

My use case is a mail server. I have an extra IP allocated to the server to use for sending e-mails so that forward and backwards DNS entries can match. The other IP address is used for HTTP virtual hosting and has many different DNS entries. Additionally using an extra IP for a dedicated mail server is in general a good practice.

Tried with Docker 1.12.5 on Linux (Ubuntu 16.04.1 LTS) with 4.8.0 kernel.

### LINKED ISSUE #35082: [SWARM] Very poor performance for ingress network with lots of parallel requests
**Description**

Executing a large number of parallel connections against plain Docker and Docker Swarms leads to 2 completely different perfomance results, with Swarm being the slowest one by a **50x** factor!
The test is reproducible (at least on my VMs) easily with Siege and the official Nginx image, but I'm actually experiencing the problem in production with our custom java-based HTTP microservice. I cannot see any obvious error message in Docker logs or kernel logs.

**Steps to reproduce the issue:**
Run the nginx container:
```
[root@stresstest01 ~]# docker run -d --rm --net bridge -m 0b  -p 80:80  --name test nginx
35c231e361d7e5ca73fb1bcfbeeaf57a066da057b708055477855e6d16af575d
```

Siege the container, and the results are good, over 13k trans/sec, and CPU in `stresstest01` is 100% used by the nginx process.

```
[root@siege01 ~]# siege -b -c 250 -t 20s -f test_vm_docker.txt >/dev/null
** SIEGE 4.0.2
** Preparing 250 concurrent users for battle.
The server is now under siege...

Lifting the server siege...
Transactions:		      260810 hits
Availability:		      100.00 %
Elapsed time:		       19.03 secs
Data transferred:	      140.03 MB
Response time:		        0.02 secs
Transaction rate:	    13705.20 trans/sec
Throughput:		        7.36 MB/sec
Concurrency:		      245.51
Successful transactions:      231942
Failed transactions:	           0
Longest transaction:	        7.03
Shortest transaction:	        0.00

```

Now, lets try with Docker Swarm (1 node swarm, 1 container stack)
```
[root@stresstest01 ~]# cat docker-compose.yml 
services:
  server:
    deploy:
      replicas: 1
    image: nginx:latest
    ports:
    - published: 80
      target: 80
version: '3.3'
```
```
[root@stresstest01 ~]# docker stack deploy test --compose-file docker-compose.yml 
Creating network test_default
Creating service test_server
```
After the first run, the results are already far worse than with plain Docker, but after the second it's
just a disaster :( Moreover, host CPU is only slightly used, and only by the nginx process. No docker related processes (`dockerd`,`cointanerd` etc) seem to be contained by CPU.

```
[root@siege01 ~]# siege -b -c 250 -t 20s -f test_vm_docker.txt >/dev/null
** SIEGE 4.0.2
** Preparing 250 concurrent users for battle.
The server is now under siege...

Lifting the server siege...
Transactions:		       65647 hits
Availability:		      100.00 %
Elapsed time:		       19.44 secs
Data transferred:	       35.28 MB
Response time:		        0.07 secs
Transaction rate:	     3376.90 trans/sec
Throughput:		        1.81 MB/sec
Concurrency:		      246.66
Successful transactions:       58469
Failed transactions:	           0
Longest transaction:	        3.02
Shortest transaction:	        0.00
 
[root@siege01 ~]# siege -b -c 250 -t 20s -f test_vm_docker.txt >/dev/null
** SIEGE 4.0.2
** Preparing 250 concurrent users for battle.
The server is now under siege...

Lifting the server siege...
Transactions:		        4791 hits
Availability:		      100.00 %
Elapsed time:		       19.47 secs
Data transferred:	        2.59 MB
Response time:		        1.00 secs
Transaction rate:	      246.07 trans/sec
Throughput:		        0.13 MB/sec
Concurrency:		      245.61
Successful transactions:        4291
Failed transactions:	           0
Longest transaction:	        1.20
Shortest transaction:	        0.00
```

**Describe the results you received:**
Good performances with plain Docker.
Very bad performances with Docker Swarm enabled.

**Describe the results you expected:**
Similar performances for the two Docker flavour on the same machine

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:41:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:49 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 17.09.0-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: s2ei2tx1nbf6lgn6d2yi9k782
 Is Manager: true
 ClusterID: s2dwwy929baleeoyk943wh2r9
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 192.168.10.187
 Manager Addresses:
  192.168.10.187:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.2.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 7.609GiB
Name: stresstest01
ID: 4XPS:KBEY:W53L:YAK6:4MZL:4HDN:DMUR:DD4T:5RWA:IUK6:522E:TCAL
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
It's a KVM virtual machine (under oVirt) but the same happens when using a physical machine.

### LINKED ISSUE #37430: Prevent bridge network driver from setting IPv4 address on bridge interface
I'm looking for a good way to prevent setting up the gateway IP address (or any other IP address, for that matter) on the bridge interface. I like the ability to configure an IP address, subnet, and default gateway on a container, but I'd like to avoid relinguishing ownership of the gateway IP itself to Docker. For instance, assume this setup before/outside of Docker:
```
# ifconfig br0
br0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
     inet 192.168.10.75  netmask 255.255.255.0  broadcast 192.168.10.255

# route -n
Destination   Gateway       Genmask        ...  Use Iface
0.0.0.0       192.168.10.1  0.0.0.0        ...    br0
192.168.10.0  0.0.0.0       255.255.255.0  ...    br0

# brctl show
bridge name   bridge id   STP enabled   interfaces
br0           XXXXXXXXX   no            eth0
                                        tap0
                                        ...
```
IOW, I have an existing bridge, with an already configured IP (or maybe not, maybe the bridge is a purely Layer-2 link), and a LAN broadcast domain with an existing subnet and default gateway. If I wanted to set up docker containers capable of fully participating on that LAN segment, including the ability to communicate with the host (if the bridge has an IPv4 address), I'd like to do the following:
```
  docker network create \
    --subnet 192.168.10.0/24 \
    --ip-range 192.168.10.192/26 \
    --gateway 192.168.10.1 \
    -o com.docker.network.bridge.name=br0 \
    docker42
```
At this point, Docker clobbers br0's IP address (.75) and replaces it with that of the gateway (.1). I can manually fix that (but shouldn't have to)...:
```
ip addr del 192.168.10.1/24 dev br0
killall dhclient
dhclient br0
```
... in order to restore br0's proper address, before bringing up a container like so:
```
docker run -d --network docker42 --ip 192.168.10.195 nginx
```
At this  point, I can connect to my container from everywhere on the LAN, including my host, and from anywhere else my LAN is reachable (e.g. via routing hops, subject to firewall rules, etc.), which is the desired result. The only obstacle is Docker's insistence on "owning" the default gateway. I want the gateway to be used as part of the IPAM setup, but HAVE TO be able to avoid having it applied to the bridge interface itself, where it doesn't belong!!!!

While using the macvlan driver instead is a possible option, I'd lose the ability to communicate with the host, and also to configure delay/loss/jitter values with e.g. "netem" on the container's host-side vethXXYYZZ interface, which is another hard requirement I have.

I'd like to introduce the "-o com.docker.network.bridge.inhibit_ipv4=true" option, which would leave the bridge interface untouched at Layer-3, but otherwise keep everything else (ipam, pool management, numbering of containers started on the given network, etc.) unchanged.

RFC pull request to follow shortly (I'm new to Docker and GoLang, so please excuse any blatant n00bness on my part :) )

### PR DESCRIPTION
**Pull request by thaJeztah**
full diff: https://github.com/docker/libnetwork/compare/feeff4f0a3fd2a2bb19cf67c826082c66ffaaed9...6659f7f4d8c1ec7a412a33f8973423ad38d22982

includes:

- docker/libnetwork#2317 Allow bridge net driver to skip IPv4 configuration of bridge interface
    - adds support for a `com.docker.network.bridge.inhibit_ipv4` label/configuration
    - fixes moby/moby#37430 Prevent bridge network driver from setting IPv4 address on bridge interface
- docker/libnetwork#2454 Support for com.docker.network.host_ipv4 driver label
    - fixes moby/moby#30053 Unable to choose outbound (external) IP for containers
- docker/libnetwork#2491 Improving load balancer performance
    - fixes moby/moby#35082 [SWARM] Very poor performance for ingress network with lots of parallel requests

**- Description for the changelog**

```markdown
- Add support for a `com.docker.network.bridge.inhibit_ipv4` driver label to skip IPv4 configuration of bridge interface docker/libnetwork#2317
- Add support for a ` com.docker.network.host_ipv4` driver label to choose outbound (external) IP for containers docker/libnetwork#2454
- Improve load balancer performance for ingress network with lots of parallel requests docker/libnetwork#2491
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
@arkodg could you have a look at;

- wether or not https://github.com/docker/libnetwork/pull/2491 should be backported to the 19.03 branch
- updating the documentation to include the new `com.docker.network.bridge.inhibit_ipv4` and `com.docker.network.host_ipv4` options
- check if any of the issues I marked as "addresses" should be changed to "fixes"

File: vendor/github.com/docker/libnetwork/vendor.conf [hunk 1]
```diff
@@ -50,6 +53,7 @@ golang.org/x/sys                        d455e41777fca6e8a5a79e34a14b8368bc11d9ba
 golang.org/x/sync                       1d60e4601c6fd243af51cc01ddf169918a5407ca
 github.com/pkg/errors                   ba968bfe8b2f7e042a574c888954fccecfa385b4 # v0.8.1
 github.com/ishidawataru/sctp            6e2cb1366111dcf547c13531e3a263a067715847
+go.opencensus.io                        9c377598961b706d1542bd2d84d538b5094d596e # v0.22.0
```
Comment (thaJeztah) [inline vendor/github.com/docker/libnetwork/vendor.conf]: This one is being updated in https://github.com/moby/moby/pull/39838

**Review by arkodg — APPROVED** [review]
LGTM thanks @thaJeztah !

**Comment by arkodg** [comment]
don't think we should back port any to 19.03
created a issue https://github.com/docker/docker.github.io/issues/10298, will need to get updated/go-live once we are closer to a 20.03 release

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
@arkodg can you check if the "addresses" should be "fixes" for all of the ones listed? I wasn't sure if they would all be fully fixed by these patches

**Comment by arkodg** [comment]
@thaJeztah  the fix should address the issues you have linked

**Comment by thaJeztah** [comment]
Thx! Changed "addresses" to "fixes"

**Comment by thaJeztah** [comment]
@AkihiroSuda @tonistiigi PTAL

**Comment by thaJeztah** [comment]
superseded by https://github.com/moby/moby/pull/40579

**Pull request closed without merging**